File: react-mcp-demo/src/main/java/com/example/reactmcp/service/ReActEventPublisher.java
package com.example.reactmcp.service;

import com.example.reactmcp.model.ReActStepEvent;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

/**
 * ReAct 事件发布器
 * 用于在 ReAct 执行过程中发布事件
 */
@Component
public class ReActEventPublisher {
    
    private final ThreadLocal<List<Consumer<ReActStepEvent>>> listeners = ThreadLocal.withInitial(ArrayList::new);
    
    /**
     * 注册事件监听器
     */
    public void registerListener(Consumer<ReActStepEvent> listener) {
        listeners.get().add(listener);
    }
    
    /**
     * 发布事件
     */
    public void publish(ReActStepEvent event) {
        List<Consumer<ReActStepEvent>> currentListeners = listeners.get();
        for (Consumer<ReActStepEvent> listener : currentListeners) {
            try {
                listener.accept(event);
            } catch (Exception e) {
                // 忽略监听器异常，避免影响主流程
            }
        }
    }
    
    /**
     * 清理当前线程的监听器
     */
    public void clear() {
        listeners.get().clear();
        listeners.remove();
    }
}

File: node-mcp-backend/src/event/reactEventPublisher.js
/**
 * ReAct 事件发布订阅器
 */
export class ReActEventPublisher {
  constructor() {
    this.listeners = [];
  }

  /**
   * 注册事件监听器
   */
  registerListener(listener) {
    this.listeners.push(listener);
  }

  /**
   * 移除事件监听器
   */
  unregisterListener(listener) {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners.splice(index, 1);
    }
  }

  /**
   * 清除所有监听器
   */
  clearListeners() {
    this.listeners = [];
  }

  /**
   * 发布事件到所有监听器
   */
  publish(event) {
    this.listeners.forEach(listener => {
      try {
        listener(event);
      } catch (error) {
        console.error('Event listener error:', error);
      }
    });
  }
}

// 全局事件发布器单例
export const globalEventPublisher = new ReActEventPublisher();
